pex/cli/commands/lock.py

```diff
@@ -119,6 +131,248 @@ def __call__(self, parser, namespace, value, option_str=None):
+                to_remove.append(existing_distribution)
+                to_install.append(distribution)
+        if retain_pip:
+            existing_distributions_by_project_name.pop(self._PIP_PROJECT_NAME, None)
+        to_remove.extend(existing_distributions_by_project_name.values())
+
+        to_unlink_by_pin = (
+            OrderedDict()
+        )  # type: OrderedDict[Tuple[ProjectName, Version], List[Text]]
+        for distribution in to_remove:
+            to_unlink = []  # type: List[Text]
+            if distribution.metadata.type is MetadataType.DIST_INFO:
+                to_unlink.extend(
+                    os.path.realpath(os.path.join(distribution.location, installed_file.path))
+                    for installed_file in Record.read(distribution.iter_metadata_lines("RECORD"))
+                )
+            elif distribution.metadata.type is MetadataType.EGG_INFO:
+                installed_files = distribution.metadata.files.metadata_file_rel_path(
+                    "installed-files.txt"
+                )
+                if installed_files:
+                    base_dir = os.path.realpath(
+                        os.path.join(distribution.location, os.path.dirname(installed_files))
+                    )
+                    to_unlink.extend(
+                        os.path.join(base_dir, path)
+                        for path in distribution.iter_metadata_lines("installed-files.txt")
+                    )
+                    # Pip generates "installed-files.txt" upon installing and this file does not
+                    # include itself; so we tack that on here ourselves.
+                    to_unlink.append(
+                        os.path.realpath(os.path.join(distribution.location, installed_files))
+                    )
+            if to_unlink:
+                to_unlink_by_pin[
+                    (distribution.metadata.project_name, distribution.metadata.version)
+                ] = [file for file in to_unlink if abs_venv_dir == commonpath((abs_venv_dir, file))]
+        if confirm and to_unlink_by_pin:
+            for (project_name, version), files in to_unlink_by_pin.items():
+                print(project_name, version, ":", file=sys.stderr)
+                for f in files:
+                    print("    ", f, file=sys.stderr)
+            if input(
+                "Remove the outdated distributions listed above from the venv at "
+                "{venv}? [yN]: ".format(venv=self.venv.venv_dir)
+            ).strip().lower() not in ("y", "yes"):
+                return Error("Sync cancelled.")
+
+        if to_unlink_by_pin:
+            parent_dirs = set()  # type: Set[Text]
+            for file in itertools.chain.from_iterable(to_unlink_by_pin.values()):
+                safe_delete(file)
+                parent_dirs.add(os.path.dirname(file))
+            for parent_dir in sorted(parent_dirs, reverse=True):
+                if not os.listdir(parent_dir) and abs_venv_dir == commonpath(
```

Comment:
Yes, if the distribution has a top-level package that is >1 level deep; i.e.: Project foo-bar 1.0 distributes code starting inside the `foo.bar` package (so `foo/bar/...` with no `foo/*.py`). I have never seen this in practice, but deleting empty dirs up to the site-packages dir would be fine. That said, that is just 1 case! There is the bin dir, and the free-form territory of the data dirs. In the bin dir case, it should never be empty - there should at least by a python executable symlink to the base python. In the data dir case I'm really unsure and a bit hesitant to go further. In general, this is terrifying territory since it strays outside of the PEX_ROOT into potentially pre-existing venvs Pex did not create.

Reply:
There are commonly-used distributions whose top-level package is >1 level deep. A noteworthy example of this is the google-cloud-* dists. They install under `site-packages/google/cloud/` where `site-packages/google/` has no files, just subdirs. 

I figure we should do whatever is consistent with the purpose of deleting the empty dirs to begin with. As far as I can tell an empty dir should have no effect (an implicit namespace package must contain some `.py` file, even if not an `__init__.py`), so I assume removing them is for hygiene/to avoid confusing people who poke around in the site-packages? In which case deleting up to the site-packages might make sense? 

But having said all that, I guess the real thing to do is whatever `pip uninstall` does in this case.  And from my experiments, it does what your code does, so no need to change anything...